exonum/src/blockchain/block.rs

```diff
@@ -349,18 +356,122 @@ impl IndexProof {
-            return Err(ProofError::AmbiguousIndex);
+            return Err(ProofError::AmbiguousEntry);
         }
-        let index_hash = *maybe_hash.ok_or(ProofError::NoIndex)?;
+        let index_hash = *maybe_hash.ok_or(ProofError::NoEntry)?;
         self.index_proof
             .check_against_hash(self.block_proof.block.state_hash)
-            .map_err(ProofError::IncorrectIndexProof)?;
+            .map_err(ProofError::IncorrectEntryProof)?;
         Ok((name.as_str(), index_hash))
     }
 }
 
+/// Proof of authenticity for a single top-level call in a block, such as a [transaction].
+///
+/// The proof consists of two parts:
+///
+/// - `block_proof`: block header with the proof of authenticity
+/// - `call_proof`: proof from the error aggregator (i.e., a `ProofMapIndex` the Merkle root
+///   of which is recorded in the block header as `error_hash`).
+///
+/// For an execution that resulted in an error, `call_proof` will be a proof of existence.
+/// If a transaction was executed successfully, `call_proof` will be a proof of absence.
+/// Since the number of transactions in a block is mentioned in the block header, the user
+/// will be able to distinguish absence of error (meaning successful execution) from
+/// the absence of a transaction with such an index. Indeed, if the index is less
+/// than amount of transactions in block, the proof denotes successful execution;
+/// otherwise, the transaction with the given index does not exist in the block.
+///
+/// Similarly, execution errors of the `before_transactions` / `after_transactions`
+/// hooks can be proven to external clients. Discerning successful execution
+/// from a non-existing service requires prior knowledge though.
+///
+/// `CallProof`s should not be confused with a proof of transaction commitment.
+/// To verify that a certain transaction was committed, use a proof from
+/// the `block_transactions` index of the [core schema].
+///
+/// [transaction]: ../runtime/struct.AnyTx.html
+/// [core schema]: struct.Schema.html
+/// [`ExecutionError`]: ../runtime/struct.ExecutionError.html
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CallProof {
+    /// Proof of authenticity for the block header.
+    #[serde(flatten)]
+    pub block_proof: BlockProof,
+
+    /// Proof of authenticity for the call status. Must contain a single key corresponding
+    /// to the `CallInBlock` in question.
+    /// The root hash of the proof must be equal to the `error_hash` mentioned in `block_proof`.
+    pub call_proof: MapProof<CallInBlock, ExecutionError>,
+
+    /// Human-readable description of an error if the call status is erroneous.
+    /// This description is not authenticated and thus should be used for diagnostic purposes only.
+    /// If the call is successful, the error description should be `None`.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub error_description: Option<String>,
+
+    #[serde(skip, default)]
```

Comment:
Can we use non exhaustive attribute for structures?

Reply:
Sure, but this can be done in a separate PRs (e.g., #1797).